frontends/web/src/routes/account/receive/receive.tsx

```diff
@@ -15,376 +15,309 @@
+    accounts,
+    code,
+    devices,
+    deviceIDs,
+}) => {
+    const { t } = useTranslation();
+    const [verifying, setVerifying] = useState<boolean>(false);
+    const [activeIndex, setActiveIndex] = useState<number>(0);
+    // index into `availableScriptTypes`, or 0 if none are available.
+    const [addressType, setAddressType] = useState<number>(0);
+    const [addressDialog, setAddressDialog] = useState<AddressDialog>();
+    const [availableScriptTypes, setAvailableScriptTypes] = useState<accountApi.ScriptType[]>();
+    const [currentAddresses, setCurrentAddresses] = useState<accountApi.IReceiveAddress[]>();
+    const [currentAddressIndex, setCurrentAddressIndex] = useState<number>(0);
 
-    public componentDidMount() {
-        if (this.getDevice() === 'bitbox') {
-            apiGet('devices/' + this.props.deviceIDs[0] + '/has-mobile-channel').then(paired => {
-                this.setState({ paired });
-            });
-        }
-    }
+    const device = deviceIDs.length ? devices[deviceIDs[0]] : undefined;
+    const account = accounts.find(({ code: accountCode }) => accountCode === code);
 
-    public UNSAFE_componentWillMount() {
-        this.registerEvents();
-    }
+    // first array index: address types. second array index: unused addresses of that address type.
+    const receiveAddresses = useLoad(accountApi.getReceiveAddressList(code));
+    const secureOutput = useLoad(accountApi.hasSecureOutput(code));
 
-    public componentWillUnmount() {
-        this.unregisterEvents();
-    }
+    useEffect(() => {
+        const handleKeyDown = (e: KeyboardEvent) => {
+            if (e.keyCode === 27 && !verifying) {
+                route(`/account/${code}`);
+            }
+        }
+        document.addEventListener('keydown', handleKeyDown);
+        return () => document.removeEventListener('keydown', handleKeyDown);
+    }, [code, verifying]);
 
-    private registerEvents = () => {
-        document.addEventListener('keydown', this.handleKeyDown);
-    }
+    // Find index in list of receive addresses that matches the given script type, or -1 if not found.
+    const getIndexOfMatchingScriptType = useCallback((
+        scriptType: accountApi.ScriptType
+    ): number => {
+        if (!receiveAddresses) {
+            return -1;
+        }
+        return receiveAddresses.findIndex(addrs => addrs.scriptType !== null && scriptType === addrs.scriptType);
+    }, [receiveAddresses]);
 
-    private unregisterEvents = () => {
-        document.removeEventListener('keydown', this.handleKeyDown);
-    }
+    useEffect(() => {
+        if (receiveAddresses) {
+            // All script types that are present in the addresses delivered by the backend. Will be empty for if there are no such addresses, e.g. in Ethereum.
+            setAvailableScriptTypes(scriptTypes.filter(sc => getIndexOfMatchingScriptType(sc) >= 0));
+        }
+    }, [receiveAddresses, getIndexOfMatchingScriptType]);
```

Comment:
what does `getIndexOfMatchingScriptType` in the deps least mean? it is a function after all.

imo everything would be easier to read if `getIndexOfMatchingScriptType` took `receiveAddresses` as an argument and it was used directly without putting it here or using `useCallback`

Reply:
Yes correct: "it is a function after all",  but if the function is not the same function it needs to execute again, therefore it needs to be in the dependency array. Or else:

> React Hook useEffect has a missing dependency: 'getIndexOfMatchingScriptType'. Either include it or remove the dependency array.eslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)

also

> The 'getIndexOfMatchingScriptType' function makes the dependencies of useEffect Hook (at line 100) change on every render. To fix this, wrap the definition of 'getIndexOfMatchingScriptType' in its own useCallback() Hook.eslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)

eslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)

Your suggestion makes sense I'll update the PR.

Reply:
changed in https://github.com/digitalbitbox/bitbox-wallet-app/pull/1668/commits/d4a2b379a27b867095f321bc367221020d4d69f3

(not yet squashed)